### DISCUSSION TITLE
Lesson 12 setup failed Custom Error

### DISCUSSION THREAD

**Discussion by DukeLurkin** [Cyfrin/foundry-full-course-cu#1356] [Q&A] [answered]
When I try to run our first unit test, "testGetUsdValue" I get this error:
"ERROR forge::runner: setUp failed reason=custom error 118cdaa7:000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266 contract=0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496"

<img width="914" alt="setupFailedCustomError92266" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/0f9d75d7-c713-4ea3-af0b-64235541d75d">

forge build compiles just fine.  

Here is my repo link:
https://github.com/DukeLurkin/foundry-defi-stablecoin-f23/tree/main

In DSCEngine.sol I've got yellow lines that say "unreachable code" here:
```solidity
s_collateralDeposited[msg.sender][
            tokenCollateralAddress
        ] += amountCollateral;
        emit CollateralDeposited(
            msg.sender,
            tokenCollateralAddress,
            amountCollateral
        );
        bool success = IERC20(tokenCollateralAddress).transferFrom(
            msg.sender,
            address(this),
            amountCollateral
        );
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }
````
I'm not sure that has anything to do with it though.  I looked through the other parts of setup and they look good to me.  Not sure what to do from here.  This started as a problem with Openzeppelin paths but I think I'm past that now???

**Comment by ghost** [comment]
My best guess is that the problem is the `ERC20Mock` from Openzeppelin, you're using the latest version which doesn't require any constructor arguments.

A couple of things you could do:
- Review your imports, check their contructor arguments and update accordingly, or
- Uninstall the library and then re-install it but with the `@v4.8.3` tag to target the same version Patrick used, like so:

```shell
forge remove lib/openzeppelin-contracts
forge install openzeppelin/openzeppelin-contracts@v4.8.3 --no-commit
```

**Reply by DukeLurkin** [reply]
I removed and installed the old OZ

<img width="974" alt="InstalledOldZep" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/69c1edc7-b514-49f7-a573-62f70bd1885a">

Changed paths that wouldn't build:

<img width="980" alt="ForgeBuildChangedPaths" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/48dc0f28-5270-4f1a-b9a8-ee4b4f6a902b">

removed ownable:

<img width="802" alt="ownableRemoved" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/04ecc400-0485-4a54-8778-79eaf07c94b9">

got it to build but with "unreachable code"

<img width="730" alt="buildUnreachableCode" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/e3565d48-bb83-4163-969c-d4b6a9422ee7">

then tried to run the test and it still won't pass:

<img width="974" alt="TestFailGetUsdValue" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/65f64838-f59b-42a9-a287-cb01b1c8285c">


I think it's still something to do with ReentrancyGuard.  

This is taking an insane amount of time to fix.

**Reply by tobezhanabi** [reply]
evm error? have you fixed this

**Reply by ghost** [reply]
@DukeLurkin push the current project to GitHub. I'll fork, review and propose a PR if I manage to make it compile.

**Reply by DukeLurkin** [reply]
Here is the repo:
https://github.com/DukeLurkin/foundry-defi-stablecoin-f23

**Reply by ghost** [reply]
@DukeLurkin 

Please check: https://github.com/DukeLurkin/foundry-defi-stablecoin-f23/pull/1

**Reply by DukeLurkin** [reply]
There is still unreachable code here: 

 s_collateralDeposited[msg.sender][tokenCollateralAddress] += amountCollateral;
        emit CollateralDeposited(msg.sender, tokenCollateralAddress, amountCollateral);
        bool success = IERC20(tokenCollateralAddress).transferFrom(msg.sender, address(this), amountCollateral);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }

I don't think I'll be able to deposit any collateral until this code is reachable.

**Reply by DukeLurkin** [reply]
I was right.  The very next thing we do is test the deposit collateral function and it can't find dsce.depositCollateral 

<img width="617" alt="depositCollateralTest" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/4e8b27db-73c2-481d-84e5-c196e2b1b397">

When you hover over the problem is: "Member "depositCollateral" not found or not visible after argument-dependent lookup in contract DSCEngine."

Go into the contract DSCEngine like it says and then you'll see it's unreachable code here:

<img width="566" alt="ureachableCodeYellowLines" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/76db3f36-e868-4342-9b5c-7e1a539bd38d">

I have no clue why this is happening.

**Reply by ghost** [reply, marked as answer]
Ok, I found the issue.

The problem is in the `isAllowedToken` modifier. The placeholder statement (`_`) is inside the if-statement  and after the `revert` statement making the deposit collateral function unreachable.

Update to this:

```diff
    modifier isAllowedToken(address token) {
        if (s_priceFeeds[token] == address(0)) {
            revert DSCEngine__NotAllowedToken();
-           _;
        }
         // this goes here, outside the if-statement
+       _;
    }
```

**Reply by DukeLurkin** [reply]
TY so much!  Is there something you did to figure that out or just experience?  Are there any steps I could have taken to know that was the issue?

**Reply by ghost** [reply]
Not much, just analyzed the `depositCollateral` function by reading the body, but when I coudn't find anything off I went into the modifiers and spot the error.

Probably the following advice doesn't apply to this error in specific, but when something throws and you cannot spot where is the issue, remember code runs synchronously and at some line the execution stops; you can place a `console.log("ran");` inside functions to verify if they are called and then you can move the console.log inside the function to find on which line the execution stops. This exercise helped me a lot at the beginning.

**Reply by DukeLurkin** [reply]
That sounds like it'll be very helpful in the future.  Thanks!
